File: lib/core/business_calendar.dart
const int normalRequestedDeliveryLeadDays = 3;
const int urgentRequestedDeliveryWindowDays = 3;

// Días de descanso obligatorio de México usados para validar días hábiles.
DateTime normalizeCalendarDate(DateTime value) {
  return DateTime(value.year, value.month, value.day);
}

bool isBusinessDay(DateTime value) {
  final date = normalizeCalendarDate(value);
  if (date.weekday == DateTime.saturday || date.weekday == DateTime.sunday) {
    return false;
  }
  return !isMandatoryRestDayInMexico(date);
}

DateTime firstAllowedNormalRequestedDeliveryDate({
  DateTime? today,
}) {
  final minimumDate = normalizeCalendarDate(
    today ?? DateTime.now(),
  ).add(const Duration(days: normalRequestedDeliveryLeadDays));
  var cursor = minimumDate;
  while (!isBusinessDay(cursor)) {
    cursor = cursor.add(const Duration(days: 1));
  }
  return cursor;
}

bool isAllowedNormalRequestedDeliveryDate(
  DateTime value, {
  DateTime? today,
}) {
  final date = normalizeCalendarDate(value);
  final minimumDate = normalizeCalendarDate(
    today ?? DateTime.now(),
  ).add(const Duration(days: normalRequestedDeliveryLeadDays));
  if (date.isBefore(minimumDate)) {
    return false;
  }
  return isBusinessDay(date);
}

bool isAllowedUrgentRequestedDeliveryDate(
  DateTime value, {
  DateTime? today,
}) {
  final date = normalizeCalendarDate(value);
  final normalizedToday = normalizeCalendarDate(today ?? DateTime.now());
  final lastAllowedDate = normalizedToday.add(
    const Duration(days: urgentRequestedDeliveryWindowDays),
  );
  if (date.isBefore(normalizedToday)) {
    return false;
  }
  return !date.isAfter(lastAllowedDate);
}

bool isSameCalendarDate(DateTime left, DateTime right) {
  return left.year == right.year &&
      left.month == right.month &&
      left.day == right.day;
}

bool isMandatoryRestDayInMexico(DateTime value) {
  final date = normalizeCalendarDate(value);

  if (_isMonthDay(date, 1, 1)) return true;
  if (_isNthWeekdayOfMonth(date, month: 2, weekday: DateTime.monday, occurrence: 1)) {
    return true;
  }
  if (_isNthWeekdayOfMonth(date, month: 3, weekday: DateTime.monday, occurrence: 3)) {
    return true;
  }
  if (_isMonthDay(date, 5, 1)) return true;
  if (_isMonthDay(date, 9, 16)) return true;
  if (_isNthWeekdayOfMonth(date, month: 11, weekday: DateTime.monday, occurrence: 3)) {
    return true;
  }
  if (_isMonthDay(date, 12, 25)) return true;
  if (_isPresidentialTransitionDay(date)) return true;

  return false;
}

bool _isMonthDay(DateTime value, int month, int day) {
  return value.month == month && value.day == day;
}

bool _isNthWeekdayOfMonth(
  DateTime value, {
  required int month,
  required int weekday,
  required int occurrence,
}) {
  if (value.month != month || value.weekday != weekday) return false;
  final ordinal = ((value.day - 1) ~/ 7) + 1;
  return ordinal == occurrence;
}

bool _isPresidentialTransitionDay(DateTime value) {
  if (value.month != 10 || value.day != 1 || value.year < 2024) {
    return false;
  }
  return (value.year - 2024) % 6 == 0;
}

File: desktop/src/renderer/lib/business-calendar.ts
export const normalRequestedDeliveryLeadDays = 3;
export const urgentRequestedDeliveryWindowDays = 3;

export function normalizeCalendarDate(value: Date) {
  return new Date(value.getFullYear(), value.getMonth(), value.getDate());
}

export function isBusinessDay(value: Date) {
  const date = normalizeCalendarDate(value);
  const weekday = date.getDay();
  if (weekday === 0 || weekday === 6) {
    return false;
  }
  return !isMandatoryRestDayInMexico(date);
}

export function firstAllowedNormalRequestedDeliveryDate(today = new Date()) {
  const minimumDate = normalizeCalendarDate(today);
  minimumDate.setDate(minimumDate.getDate() + normalRequestedDeliveryLeadDays);
  const cursor = new Date(minimumDate);
  while (!isBusinessDay(cursor)) {
    cursor.setDate(cursor.getDate() + 1);
  }
  return cursor;
}

export function isAllowedNormalRequestedDeliveryDate(value: Date, today = new Date()) {
  const date = normalizeCalendarDate(value);
  const minimumDate = normalizeCalendarDate(today);
  minimumDate.setDate(minimumDate.getDate() + normalRequestedDeliveryLeadDays);
  if (date < minimumDate) {
    return false;
  }
  return isBusinessDay(date);
}

export function isAllowedUrgentRequestedDeliveryDate(value: Date, today = new Date()) {
  const date = normalizeCalendarDate(value);
  const normalizedToday = normalizeCalendarDate(today);
  const lastAllowedDate = normalizeCalendarDate(today);
  lastAllowedDate.setDate(lastAllowedDate.getDate() + urgentRequestedDeliveryWindowDays);
  return date >= normalizedToday && date <= lastAllowedDate;
}

function isMandatoryRestDayInMexico(value: Date) {
  return (
    isMonthDay(value, 1, 1) ||
    isNthWeekdayOfMonth(value, 1, 2, 1) ||
    isNthWeekdayOfMonth(value, 1, 3, 3) ||
    isMonthDay(value, 5, 1) ||
    isMonthDay(value, 9, 16) ||
    isNthWeekdayOfMonth(value, 1, 11, 3) ||
    isMonthDay(value, 12, 25) ||
    isPresidentialTransitionDay(value)
  );
}

function isMonthDay(value: Date, month: number, day: number) {
  return value.getMonth() + 1 === month && value.getDate() === day;
}

function isNthWeekdayOfMonth(value: Date, weekday: number, month: number, occurrence: number) {
  if (value.getMonth() + 1 !== month) return false;
  if (value.getDay() !== weekday) return false;
  const ordinal = Math.floor((value.getDate() - 1) / 7) + 1;
  return ordinal === occurrence;
}

function isPresidentialTransitionDay(value: Date) {
  const year = value.getFullYear();
  return value.getMonth() + 1 === 10 && value.getDate() === 1 && year >= 2024 && (year - 2024) % 6 === 0;
}
